drivers/spi/spi_pl022.c

```diff
@@ -0,0 +1,419 @@
+	struct spi_pl022_data *data = dev->driver_data;
+	u32_t base = cfg->base;
+	u16_t value;
+
+	/* Check Rx FIFO is Not Empty */
+	while (sys_test_bit(SPI_REG_ADDR(base, SPI_SR_OFFSET), RX_NOEMPTY)) {
+		value = sys_read16(SPI_REG_ADDR(base, SPI_DR_OFFSET));
+
+		if (spi_context_rx_buf_on(&data->ctx)) {
+			UNALIGNED_PUT(value, (u8_t *)data->ctx.rx_buf);
+		}
+
+		spi_context_update_rx(&data->ctx, 1, 1);
+	}
+}
+
+static void push_data(struct device *dev)
+{
+	const struct spi_pl022_config *cfg = dev->config->config_info;
+	struct spi_pl022_data *data = dev->driver_data;
+	u32_t base = cfg->base;
+	u16_t value;
+
+	/* Check Tx FIFO is Not Full and Hold, if Rx FIFO Full */
+	if (sys_test_bit(SPI_REG_ADDR(base, SPI_SR_OFFSET), TX_NOFULL) &&
+	    !sys_test_bit(SPI_REG_ADDR(base, SPI_SR_OFFSET), RX_FULL)) {
+		value = 0;
+
+		if (spi_context_tx_buf_on(&data->ctx)) {
+			value = UNALIGNED_GET((u8_t *)(data->ctx.tx_buf));
+		}
+
+		sys_write16(value, SPI_REG_ADDR(base, SPI_DR_OFFSET));
+		spi_context_update_tx(&data->ctx, 1, 1);
+	}
+
+	/* Disabling Interrupt when no data to Transmit */
+	if (!spi_context_tx_on(&data->ctx)) {
+		sys_clear_bit(SPI_REG_ADDR(base, SPI_IMSC_OFFSET), SPI_TXMIS);
+	}
+}
+
+static void set_frequency(struct device *dev, u32_t freq)
+{
+	const struct spi_pl022_config *cfg = dev->config->config_info;
+	u32_t base = cfg->base;
+	u32_t prescale;
+	u32_t clk_rate;
+	u32_t spi_clock;
+
+	/* Get periperal clock frequency */
+	clock_control_get_rate(device_get_binding(DT_CLOCK_CONTROL_LABEL),
+			       (clock_control_subsys_t) &cfg->pclk, &spi_clock);
+
+	for (prescale = CLK_PRESCALE_MIN;
+	     prescale <= CLK_PRESCALE_MAX; prescale += 2) {
+		for (clk_rate = SERIAL_CLKRATE_MIN;
+		     clk_rate <= SERIAL_CLKRATE_MAX; clk_rate++) {
+			if (freq >= (spi_clock / (prescale * clk_rate))) {
+				break;
+			}
+		}
+	}
+
+	/* Set prescale */
+	sys_set_bits(SPI_REG_ADDR(base, SPI_CPSR_OFFSET), BYTE_MASK,
+		     CLK_PRESCALE_SHIFT, prescale);
+
+	/* Set clkrate */
+	sys_set_bits(SPI_REG_ADDR(base, SPI_CR0_OFFSET), BYTE_MASK,
+		     SERIAL_CLKRATE_SHIFT, clk_rate);
+}
+
+static void completed(struct device *dev, u32_t error)
+{
+	const struct spi_pl022_config *cfg = dev->config->config_info;
+	struct spi_pl022_data *data = dev->driver_data;
+	u32_t base = cfg->base;
+
+	if (error) {
+		goto out;
+	}
+
+	if (spi_transfer_complete(dev)) {
+		return;
+	}
+out:
+	/* Disabling Interrupts */
+	sys_write32(0, SPI_REG_ADDR(base, SPI_IMSC_OFFSET));
+
+	/* Disabling SSP */
+	sys_clear_bit(SPI_REG_ADDR(base, SPI_CR1_OFFSET), SSE);
+
+	spi_context_cs_control(&data->ctx, false);
+
+	LOG_DBG("SPI transaction complete %s error",
+		    error ? "with" : "without");
+
+	spi_context_complete(&data->ctx, error ? -EIO : 0);
+}
+
+static void spi_pl022_isr(struct device *dev)
+{
+	const struct spi_pl022_config *cfg = dev->config->config_info;
+	u32_t base = cfg->base;
+	u32_t error = 0;
+	u32_t status;
+
+	LOG_DBG("%p", dev);
```

Comment:
Not a very useful log message, to be retained in the code.

Reply:
removed